File: snowfall/render.c
//src/render.c

#include "render.h"

#include <raylib.h>

#include "grid.h"

static Color SNOWFLAKE_COLOUR_TO_RAYLIB_COLOUR[] = {BLACK, WHITE};

void snowfallInitWindow() {
    InitWindow(WINDOW_WIDTH, WINDOW_HEIGHT, WINDOW_TITLE);
    SetTargetFPS(WINDOW_FPS);
}

void snowfallDestroyWindow() {
    CloseWindow();
}

static void drawSingleSnowflake(const Grid* grid, const int64_t row, const int64_t col) {
    const int64_t idx = getGridIndex(col, row);
    DrawRectangle((int) col * SNOWFALL_CELL_SIZE,
        (int) row * SNOWFALL_CELL_SIZE,
        SNOWFALL_CELL_SIZE,
        SNOWFALL_CELL_SIZE,
        SNOWFLAKE_COLOUR_TO_RAYLIB_COLOUR[grid->colours[idx]]);
}

static void drawSnowflakes(const Grid* grid) {
    for (int64_t row = 0; row < SNOWFALL_GRID_HEIGHT; row++) {
        for (int64_t col = 0; col < SNOWFALL_GRID_WIDTH; col++) {
            drawSingleSnowflake(grid, row, col);
        }
    }
}

bool snowfallRenderReportWindowClose(const Grid* grid) {
    BeginDrawing();
    ClearBackground(SKYBLUE);
    drawSnowflakes(grid);
    EndDrawing();
    return WindowShouldClose();
}

File: snowfall_rust/src/render.rs
//src/render.rs

use macroquad::prelude::*;
use crate::constants::{CELL_SIZE, GRID_HEIGHT, GRID_WIDTH};
use crate::sim::{get_grid_index, CellState};
use crate::sim::CellState::Snow;

fn draw_single_snowflake(col: usize, row: usize) {
    draw_rectangle((col * CELL_SIZE) as f32,
                   (row * CELL_SIZE) as f32,
                   CELL_SIZE as f32,
                   CELL_SIZE as f32,
                   WHITE
    )
}

pub async fn render(grid: &mut Box<[CellState]>) {
    request_new_screen_size((GRID_WIDTH * CELL_SIZE) as f32,
                            (GRID_HEIGHT * CELL_SIZE) as f32);
    clear_background(BLACK);

    for row in 0..GRID_HEIGHT {
        for col in 0..GRID_WIDTH {
            if grid[get_grid_index(col, row)] == Snow {
                draw_single_snowflake(col, row);
            }
        }
    }

    next_frame().await;
}